pkg/sr/mock/registry.go

```diff
@@ -0,0 +1,650 @@
+	if err = r.validateReferences(schema); err != nil {
+		return 0, 0, err
+	}
+
+	// Get or create subject data
+	subj := r.getOrCreateSubject(subject)
+
+	// Check if subject is soft-deleted
+	if subj.isDeleted {
+		return 0, 0, errSubjectNotFound(subject)
+	}
+
+	// existing identical schema under same subject?
+	for v, versionData := range subj.versions {
+		if !versionData.isDeleted && r.schemasEqual(versionData.schema.Schema, schema) {
+			return versionData.schema.ID, v, nil
+		}
+	}
+
+	// reuse id from other subject?
+	for existingID, existing := range r.schemasByID {
+		if !r.schemasEqual(existing, schema) {
+			continue
+		}
+		version = subj.highestVersion + 1
+
+		// Check for circular dependencies before storing the schema
+		if err = r.validateCircularReferences(schema, subject, version); err != nil {
+			return 0, 0, err
+		}
+
+		// Store schema version data
+		subj.versions[version] = &versionData{
+			schema: sr.SubjectSchema{
+				Subject: subject,
+				Version: version,
+				ID:      existingID,
+				Schema:  schema,
+			},
+			isDeleted: false,
+		}
+		subj.latestVersion = version
+		subj.highestVersion = version
+		return existingID, version, nil
+	}
+
+	// brand-new schema
+	id = r.nextID
+	r.nextID++
+	version = subj.highestVersion + 1
+
+	// Check for circular dependencies before storing the schema
+	if err = r.validateCircularReferences(schema, subject, version); err != nil {
+		return 0, 0, err
+	}
+
+	// Store schema globally
+	r.schemasByID[id] = schema
+
+	// Store schema version data
+	subj.versions[version] = &versionData{
+		schema: sr.SubjectSchema{
+			Subject: subject,
+			Version: version,
+			ID:      id,
+			Schema:  schema,
+		},
+		isDeleted: false,
+	}
+	subj.latestVersion = version
+	subj.highestVersion = version
+	return id, version, nil
+}
+
+// SetCompat emulates PUT /config/{subject}.
+func (r *Registry) SetCompat(subject string, c sr.CompatibilityLevel) error {
+	if c.String() == "" {
+		return fmt.Errorf("empty compatibility level")
+	}
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	// Get or create subject data
+	subj := r.getOrCreateSubject(subject)
+
+	subj.compatLevel = &c
+	return nil
+}
+
+// GetSchema retrieves a schema by subject and version.
+func (r *Registry) GetSchema(subject string, version int) (sr.SubjectSchema, bool) {
+	return r.getSchemaBySubjectVersion(subject, version)
+}
+
+// SubjectExists returns true if subject exists and has not been hard-deleted.
+func (r *Registry) SubjectExists(subject string) bool {
+	r.mu.RLock()
+	defer r.mu.RUnlock()
+
+	subj, ok := r.subjects[subject]
+	return ok && !subj.isDeleted
+}
+
+// Reset wipes all in-memory state. Useful between tests.
+func (r *Registry) Reset() {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	r.subjects = make(map[string]*subjectData)
+	r.schemasByID = make(map[int]sr.Schema)
+	r.nextID = 1
+	r.globalCompat = sr.CompatBackward
+}
+
+// getSchemaBySubjectVersionUnsafe is like getSchemaBySubjectVersion but assumes caller holds lock
```

Comment:
I usually reserve unsafe for things that literally use `unsafe` and would swap Unsafe for Locked, but this works fine too.